p2panda-engine/src/operation.rs

```diff
@@ -0,0 +1,260 @@
+        ciborium::from_reader::<Header<E>, _>(header).map_err(Into::<DecodeError>::into)?;
+    let body = body.map(Body::new);
+    Ok((header, body))
+}
+
+#[derive(Debug, Error)]
+pub enum DecodeError {
+    /// An error occurred while reading bytes
+    ///
+    /// Contains the underlying error returned while reading.
+    #[error("an error occurred while reading bytes: {0}")]
+    Io(std::io::Error),
+
+    /// An error occurred while parsing bytes
+    ///
+    /// Contains the offset into the stream where the syntax error occurred.
+    #[error("an error occurred while parsing bytes at position {0}")]
+    Syntax(usize),
+
+    /// An error occurred while processing a parsed value
+    ///
+    /// Contains a description of the error that occurred and (optionally) the offset into the
+    /// stream indicating the start of the item being processed when the error occurred.
+    #[error("an error occurred while processing a parsed value at position {0:?}: {1}")]
+    Semantic(Option<usize>, String),
+
+    /// The input caused serde to recurse too much
+    ///
+    /// This error prevents a stack overflow.
+    #[error("recursion limit exceeded while decoding")]
+    RecursionLimitExceeded,
+}
+
+impl From<CiboriumError<std::io::Error>> for DecodeError {
+    fn from(value: CiboriumError<std::io::Error>) -> Self {
+        match value {
+            CiboriumError::Io(err) => DecodeError::Io(err),
+            CiboriumError::Syntax(offset) => DecodeError::Syntax(offset),
+            CiboriumError::Semantic(offset, description) => {
+                DecodeError::Semantic(offset, description)
+            }
+            CiboriumError::RecursionLimitExceeded => DecodeError::RecursionLimitExceeded,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum IngestResult<E> {
+    /// Operation has been successfully validated and persisted.
+    Complete(Operation<E>),
+
+    /// We're missing previous operations before we can try validating the backlink of this
+    /// operation.
+    ///
+    /// The number indicates how many operations we are lacking before we can attempt validation
+    /// again.
+    Retry(Header<E>, Option<Body>, u64),
+}
+
+/// Checks an incoming operation for log integrity and persists it into the store when valid.
+///
+/// This method also automatically prunes the log when a prune flag was set in the header.
+///
+/// If the operation seems valid but we're still lacking information (as it might have arrived
+/// out-of-order) this method does not fail but indicates that we might have to retry again later.
+///
+/// The trait bounds requires the operation header to contain a prune flag as specified by the core
+/// p2panda specification.
+pub async fn ingest_operation<S, L, E>(
+    store: &mut S,
+    header: Header<E>,
+    body: Option<Body>,
+) -> Result<IngestResult<E>, IngestError>
+where
+    S: OperationStore<L, E> + LogStore<L, E>,
+    E: Clone + Serialize + DeserializeOwned + Extension<L> + Extension<PruneFlag>,
```

Comment:
Love our extensions :wink:

Reply:
Although.... we could also move at least `log_id` into the method signature, and not require that `Extension<L>` be present. This gives users more flexibility in how they define log ids, it doesn't need to be something encoded in the operation extensions. The same could be said for prune flag even... but there it gets a bit fuzzy, as we specify where pruning should happen at a protocol level, and to respect that we say a published prune flag is required.